cats/cats-sql/src/main/kotlin/com/netflix/spinnaker/cats/sql/cluster/SqlClusteredAgentScheduler.kt

```diff
@@ -241,7 +238,16 @@ class SqlClusteredAgentScheduler(
       .forEach {
         log.warn("Found zombie agent {}, removing it", it.key)
         activeAgents.remove(it.key, it.value)
-    }
+        // Cancel zombie futures interrupting their AgentExecutionAction threads
+        val f = activeAgentsFutures.remove(it.key)
+        if (f == null) {
+          log.warn("Agent execution without future for cancelling it: {}", it.key)
+        } else {
+          if (!f.cancel(true) && !f.isDone) {
```

Comment:
Sorry for not catching this earlier. 
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/Future.html#cancel(boolean) says:
```
After this method returns, subsequent calls to isDone() will always return true
```
Doesn't this mean the if condition will never be true?

Reply:
Haha that's right, so how about this
```
if (!f.cancel(true) && !f.isCancelled) {
            log.error("Unable to cancel execution for agent {} after {}ms (Future.cancel returned false). This may leak resources!", it.key, zombieAgentThreshold)
}
```
We need a manual to interpret the meaning of these booleans 😄

Reply:
Trying to lay out the baby steps to see if I've got my head on straight.

I think we agree the thing that we care about is when we can't cancel the task.  And so the natural thing would be to log whenever the cancel method returns false.  Except for this little note in the javadocs:
```
false if the task could not be cancelled, typically because it has already completed normally
```
So if `cancel` returns false, and `isCancelled` is false, does that mean the task completed normally?  In that case we don't want to log.  And I guess, if `isCancelled` is true, we don't really need a log message either.

So that one case where `cancel` returns false, `isCancelled` is false, and the task didn't complete normally -- that's when we want to log...but I'm not sure we have enough information to determine that.  But maybe the log message you've got (the part about `may leak resources`) is sufficiently vague to be just right :)